\chapter{Introduction}


In this work we will focus on the methods that are used for automatic
comparative analysis of biological sequences and annotation of the biological
sequences. We will study the methods, their complexity, possible improvements
and propose new methods that will produce alignments and annotations with higher
quality. Since the new technologies are giving ability to produce more and more
data, the automatic analysis of such data is necessary and therefore it is
important to develop such tools.

By biological sequences we mean either DNA/RNA sequences, which are finite
sequences over the alphabet $\{A,C,G,T/U\}$ ($T$ is in DNA, $U$ in RNA) or the
protein sequences which are sequences over the alphabet of $20$ amino acids.
Later in this chapter we will discuss the relationship between those sequences.
We focus on two problems. The first is the annotation of the biological
sequences and the second is the comparative analysis of such sequences. For both 
problems we will use the variants of the \abbreviation{hidden Markov
models}{HMMs}.

Hidden Markov models are generative probabilistic models which define the
probability distribution of the sequences with their annotation (formally HMMs
generate pair of sequence with annotation) \cite{Durbin1998}. Annotation is
the assignments of the labels to the sequence according it's meaning (function,
the origin of the part of the sequence or any other information). HMMs are used
for the probabilistic analysis of the sequences of various fields. In our work
we focus on  the biological sequences. 

Decoding algorithm is algorithm that from the input sequence $X$ and HMM $H$
computes the annotation that could generate $X$. Goal is to recover the true
annotation that generated $X$, but finding such annotation is impossible due to
the randomness in the generating process. Therefore the decoding algorithm finds
some approximation of the true annotation.  In recent years it has beed nown
that the different decoding methods can used to improve the ``quality'' of the
annotations of sequences
\cite{Gross2007,Nanasi2010,Nanasi2010mgr,Truszkowski2011}.  Additionally there
are many techniques that are used to lower the computations complexity of the
decoding and training algorithms (training algorithm are algorithm that computes
the parameters of the model from the training data set).

In the field of the comparative analysis of a biological sequence we focus on the
sequence alignment problem. Sequence alignment is simple data format (or data
structure) that represent the similarities in the sequences. Alignment is
created by inserting dashes ('-') into the in a way that both sequences has same
lengths and similar symbols are on the same positions in both sequences. In
computational biology the alignments are constructed in a way that the
homologous parts of the sequences are aligned together (they are on the same
positions) and parts that are not homologous to the other sequence are aligned
to a gap symbols ('-'). Homologous sequences are sequences that have evolved
from the same part of the sequence of the ancestral sequence (we assume that
current organisms evolve from single ancestral organism through series of 
mutations, insertions, deletion and other transformations).
Non-homologous parts corresponds to the parts that were either inserted into 
the sequence during evolution or that their's homologous part in the second
organism was deleted during it's evolution.

The reason for searching and studying homologous sequences is that homologous
sequences are expected to have same or similar functions in both organisms.
Therefore by studying homologous sequences we can transfer knowledge about
functions from one organisms to other organisms. We focus on the methods for
constructing alignments of the biological sequences, not on the actual
biological research.

Variant of the hidden Markov models called \abbreviation{pair hidden Markov
models}{pHMMs} can be used to create sequence alignments. Unlike HMM, pHMM is
probabilistic generative model that generate three objects: the sequences $X,Y$
and their alignment (or annotation) $A$. Similarly as with the HMMs, the
sequences $X$ and $Y$ are observed but the alignment $A$ is hidden. By proper
decoding method we can obtain the good approximation of $A$. Our goal is to
study decoding methods and models that are used for sequence alignment in order
to create software that will produce alignments with high quality.

\section{Structure}

The structure of this document is following. Later in this chapter we describe
necessary biological term that will be used later and define several formal
notations that will be used later.

In the second chapter we will discuss the sequence alignments and the methods
that are used to create sequences alignment without using pair hidden Markov
models. We describe basic scoring schemes, algorithms that search for alignments
with high score and the several approaches how to improve the computational
complexity of such algorithms.

The third chapter discuss hidden Markov models. We define HMMs, review basic
algorithm that are used with HMMs. We discuss the highest expected gain
framework which is generalization of the decoding algorithms. Later we define
several variants of HMMs: generalized hidden Markov models, high-order HMMs,
pair hidden Markov models and their combinations. In the end of chapter we
discuss several algorithmic improvements of algorithms that are used with HMMs
that reduces the computational complexity.

Fourth chapter is dedicated to the methods for pairwise alignment (alignment of
two sequences) using pair hidden Markov models and their variants and also to
other applications of pHMMs. We review several pairwise aligners and comparative
gene finders and also results concerning the quality and biases in the
alignments.

The fifth chapter contain our PhD project with the proposed directions for our
research. We want to focus on the improvements and the analysis of the decoding
algorithms and their application to the sequence alignments. 

\section{Biological Introduction}

%\todo{zmen nazov kapitoly}

%DNA,RNA,protein, amino-acid, residue, base, gene -- exon, conod, start codon,
%stop codon, cDNA, intron, splice site, donor, acceptor, intergenic region,
%homolog,

In review several biological terms that will be needed. The more information
about DNA, proteins and genes can be found in
\cite{BiologyForDummies,UnderstandingBioinformatics}.
Every cell of the living organisms contains \firstUseOf{DNA} molecule. DNA is
double stranded molecule, it consists of two long sequences of
\firstUseOf{nucleotides}. There are four types of nucleotides in DNA: adenosine,
cytosine, guanine and thymine represented by letters $A,C,G$ and $T$
respectively (In RNA the nucleotide $T$ is replaced with uracil, denoted by
$U$). The strands in DNA are complementary, the nucleotides on the same position
in strands are connected with hydrogen bonds and are complementary: $A$ is
always connected with $T$ and $C$ is always connected with $C$ by hydrogen
bonds. Therefore we can represent DNA by one sequence aver alphabet
$\{A,C,G,T\}$ (the complementary strand can be easily computed).

In DNA are encoded \firstUseOf{proteins}. Proteins play important role in the
cell biology since  they regulates many processes in the cell and are a
catalysts to many chemical reactions. Proteins are sequences of
\firstUseOf{amino-acid} molecules. There are $20$ amino-acids that can be
encoded in DNA. Parts of the DNA that encodes proteins are called genes (gene
is ``substring'' of DNA that will be translated into one protein).

In the following text we describe the structure of eucaryotic genes, which will be
needed in following chapters. Every gene basically consists from two types of
sequences: exons which encods amino-acids, and introns which will be removed
from gene before translation (gene does not have to had intron). At first,
double stranded DNA is split into single stranded DNA. Once gene is isolated (in
form of mRNA), the introns are spliced out and exons are merged into one
sequence: spliced-mRNA. This sequence is then translated into the sequence of
amino-acids (protein). Each amino acid is encoded by triplet of nucleotides
called \firstUseOf{codons}. Amino-acids have multiple representation: different
codons can encode same amino-acid. Since codons are always triplets, the length
of spliced-mRNA is multiple of three. However, the length of the exons does not
have to be multiple of three since introns can be placed inside codons.  The
sequence that gene begins with is called start codon (the sequence AUG) and the
last sequence of a gene is called stop codon (the sequences $TAG,TAA$ or $TGA$).

The structure of intron is following. Intron can be divided into five parts: it
begins with (also called donor site), intron, branch point, branch-acceptor and
ends with 5' splice site  3' splice site (or splice acceptor site).  Acceptor
site, branch point and donor have fixed length
\cite{Pairagon2009,UnderstandingBioinformatics}.


\section{Notations}

In this section we summarize the notations that we will use in the following
chapters.

All sequences, members of sets, vectors of matrices will be indexed from $0$
(the first element of a sequence is $0$-th elements). We will use mostly
the right-open intervals: $I=[a,b)$ means the $a\in I$ but $b\notin I$. 

Element on the $k$-th position (zero based) of the string $s$
will be written as $s[k]$. Substring beginning at position $i$ (including) and
ending at position $j$ (excluding symbol on position $j$) is $s[i:j]$.  If $n$
is the length of the sequence then $s[:i]$ is equivalent to $s[0:i]$ and $s[i:]$
is equivalent to $s[i:n]$.  We will use the terms sequence and string
interchangeably since they have same meanings. The term sequence is often used
in the computational biology so also we will use it.

Let $M$ be the matrix. Then $M[i,j]$ is the element from the $i$-th row and
$j$-th column of $M$ (indices are zero based). Similarly as to strings,
submatrix $M[i:j,k:l]$ is matrix consisting from the intersection of rows
$i,i+1,\dots, j-1$ with columns $k,k+1,\dots,l-1$. Additionally, if $M$ is of
size $n\times m$ then $M[:i,j:]$ is equivalent to $M[0:i,j:m]$.  The term
$M[i,:]$ is equivalent to $M[i,0:m]$ which is the $i$-th row of $M$.
